### PR TITLE
New Element::getElementByPath function.

### PR DESCRIPTION
**Pull request by feldstj**
As pointed out by Bernard, perhaps the function should be renamed from getElementByPath to something else. Bernard recommended getChildByPath or just extending getChild, my only concern with this is that the result could be a descendant that is more than one level deep, so technically not a child, but a descendant. Thoughts appreciated.

### CODE REVIEW

File: source/MaterialXCore/Element.h [hunk 1]
```diff
@@ -117,6 +117,13 @@ class Element : public std::enable_shared_from_this<Element>
     ///    the returned path will be relative to this ancestor.
     string getNamePath(ConstElementPtr relativeTo = nullptr) const;
 
+    /// Return the element referred to by the path relative to the current
+    /// element. A path of empty string returns a pointer to the current
+    /// element.
+    /// @param path The path to use to find an element relative to the
```
Comment (bernardkwok) [inline source/MaterialXCore/Element.h]: You should describe what happens if the path is the Elements name or if the path is not found. I believe that you also end up returning the Element. So "failure" I assume always menas to return itself?

File: source/MaterialXCore/Element.h [hunk 2]
```diff
@@ -117,6 +117,13 @@ class Element : public std::enable_shared_from_this<Element>
...
+    ///    current element.
+    ElementPtr getElementByPath(const std::string& path);
```
Comment (bernardkwok) [inline source/MaterialXCore/Element.h]: As already discussed this sounds more like getting a child or a descendent of this element which seems less ambiguous. Also should this method not be const since it doesn't modify it's own contents?

File: source/MaterialXTest/Document.cpp [hunk 3]
```diff
@@ -32,6 +32,13 @@ TEST_CASE("Document", "[document]")
     REQUIRE(constant->getNamePath() == "nodegraph1/node1");
     REQUIRE(constant->getNamePath(nodeGraph) == "node1");
 
+    // Test getting elements by path
+    REQUIRE(doc->getElementByPath("") == doc);
+    REQUIRE(doc->getElementByPath("nodegraph1") == nodeGraph);
+    REQUIRE(doc->getElementByPath("nodegraph1/node1") == constant);
+    REQUIRE(nodeGraph->getElementByPath("") == nodeGraph);
+    REQUIRE(nodeGraph->getElementByPath("node1") == constant);
```
Comment (bernardkwok) [inline source/MaterialXTest/Document.cpp]: Should probably add a test with a non-empty path which is not found.

**Comment by bernardkwok** [comment]
General comment: This change will need to be against adsk_update and not adsk_update_shadergen for review to be considered to merge with the ILM repo.

File: source/MaterialXCore/Element.h (hunk 2 above)
Comment (feldstj) [reply source/MaterialXCore/Element.h]: It's returning an ElementPtr though, not a ConstElementPtr, so it can't be const.

File: source/MaterialXTest/Document.cpp (hunk 3 above)
Comment (feldstj) [reply source/MaterialXTest/Document.cpp]: Sure, I can add that.

Comment (feldstj) [reply source/MaterialXTest/Document.cpp]: And for the comment below (since I can't reply to it directly), I can add this to both repos.

File: source/MaterialXCore/Element.h (hunk 1 above)
Comment (feldstj) [reply source/MaterialXCore/Element.h]: No, if the path is not found, then I believe ElementPtr() is returned.

File: source/MaterialXCore/Element.h (hunk 2 above)
Comment (niklasharrysson) [reply source/MaterialXCore/Element.h]: I agree "child" would be better, and then to also change the behavior above to return nullptr instead of itself if a matching child is not found. In order to follow the existing getNamePath method maybe it could be named getChildByNamePath in that case?

File: source/MaterialXCore/Element.h (hunk 1 above)
Comment (niklasharrysson) [reply source/MaterialXCore/Element.h]: +1, If no exact match if found it would be better to return nullptr.

File: source/MaterialXCore/Element.cpp [hunk 4]
```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
...
+    ElementPtr currentElement = getSelf();
+    if (elementNames.size() == 0)
+    {
+        return currentElement;
+    }
+    for (std::string elementName : elementNames)
+    {
+        if (!(currentElement = currentElement->getChild(elementName)))
+        {
+            return currentElement;
```
Comment (niklasharrysson) [inline source/MaterialXCore/Element.cpp]: If a path is given that only partially matches the hierarchy this will return the element at that sub-path right, and ignore the rest of the path? Would it not be better to require an exact match to the leaf element in the given path, and return nullptr if it's not found?

Comment (niklasharrysson) [reply source/MaterialXCore/Element.cpp]: Sorry my bad, currentElement is null in that case. But instead of returning currentElement you could explicitly return ElementPtr() to make the code more readable :)

Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: I don't think that's what would happen. If currentElement->getChild(elementName) fails, then this returns ElementPtr().

Comment (niklasharrysson) [reply source/MaterialXCore/Element.cpp]: Yep, that's right (as I wrote in my correction above), but I would still return ElementPtr() to make that more clear.

Here's a version with my suggested changes:

ElementPtr Element::getChildByNamePath(const std::string& path)
{
    const std::vector<std::string> elementNames = splitString(path, NAME_PATH_SEPARATOR);
    if (elementNames.empty())
    {
        return ElementPtr();
    }
    ElementPtr currentElement = getSelf();
    for (const std::string& elementName : elementNames)
    {
        if (!(currentElement = currentElement->getChild(elementName)))
        {
            return ElementPtr();
        }
    }
    return currentElement;
}

Note that empty string is now returning null instead of self, which feels more correct if it's a child element you're requesting.
Just my 2 cents.

Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: I'm fine with being more explicit and returning ElementPtr() instead of currentElement when the condition fails. I can see how people might miss that that's what it is doing.

For my purposes, I really want to get a Document when a user specifies document->getElementByPath(""); I agree it doesn't match the getChildByNamePath name very well though, but I didn't originally call it that. :) It makes sense that this would work though as document->getNamePath() would return "" and I'm just doing the opposite with document->getElementByPath("")

Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: Seems reasonable, I'll make the above suggestions soon.

File: source/MaterialXCore/Element.h (hunk 1 above)
Comment (bernardkwok) [reply source/MaterialXCore/Element.h]: Okay. Still missing a comment if the path if not found what is returned.
Also, I think the English is a bit funny. Maybe something like: "If an empty string is provided as the path then a shared pointer to the current element is returned. If the path cannot be found then an empty shared pointer is returned" ?

File: source/MaterialXCore/Element.cpp [hunk 5]
```diff
@@ -116,7 +116,7 @@ ElementPtr Element::getElementByPath(const std::string& path)
     {
```
Comment (niklasharrysson) [inline source/MaterialXCore/Element.cpp]: Nitpick: The elementName could be a const reference to avoid copying the string. The elementNames vector could also be declared const above since it never changes.

File: source/MaterialXCore/Element.h [hunk 6]
```diff
@@ -117,6 +117,13 @@ class Element : public std::enable_shared_from_this<Element>
     ///    the returned path will be relative to this ancestor.
     string getNamePath(ConstElementPtr relativeTo = nullptr) const;
 
+    /// Return the element referred to by the path relative to the current
+    /// element. A path of empty string returns a pointer to the current
+    /// element.
```
Comment (niklasharrysson) [inline source/MaterialXCore/Element.h]: Nitpick: "the path relative to" could be rephrased to "the hierarchical name path relative to" to be even more clear on what the path is.

File: source/MaterialXCore/Element.cpp (hunk 5 above)
Comment (niklasharrysson) [reply source/MaterialXCore/Element.cpp]: elementName is still not a reference, the & is missing. Otherwise it looks good, I'm approving now.

Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: whoops, will fix.

Comment (niklasharrysson) [reply source/MaterialXCore/Element.cpp]: thanks

File: source/MaterialXCore/Element.cpp [hunk 7]
```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
     return res;
 }
 
+ElementPtr Element::getElementByNamePath(const std::string& path)
+{
+    const std::vector<std::string> elementNames = splitString(path, NAME_PATH_SEPARATOR);
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Element.cpp]: Just a minor note that you can omit the "std::" qualifier for string and vector within the MaterialX namespace.

File: source/MaterialXCore/Element.cpp [hunk 8]
```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
...
+    ElementPtr currentElement = getSelf();
+    if (elementNames.size() == 0)
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Element.cpp]: Perhaps better as "if (elementNames.empty())"?

File: source/MaterialXCore/Element.cpp [hunk 9]
```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
     return res;
 }
 
+ElementPtr Element::getElementByNamePath(const std::string& path)
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Element.cpp]: Since this method can only find descendants of the given element, what would you think of the method name "getDescendant"?  To my mind, the name path is the only input that can uniquely qualify an element's descendant, so perhaps the "byNamePath" extension is unneeded.

Comment (bernardkwok) [reply source/MaterialXCore/Element.cpp]: I had voted for using "descendent" as well :)

File: source/MaterialXCore/Element.cpp (hunk 7 above)
Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: Ah, ok, didn't realize that.

File: source/MaterialXCore/Element.cpp (hunk 8 above)
Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: Good suggestion.

File: source/MaterialXCore/Element.cpp (hunk 9 above)
Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: Haha, ya, we had some discussions about this. I like "getDescendant". It's more precise.

File: source/MaterialXCore/Element.cpp [hunk 10]
```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
     return res;
 }
 
+ElementPtr Element::getDescendant(const string& path)
+{
+    const vector<string> elementNames = splitString(path, NAME_PATH_SEPARATOR);
+    ElementPtr currentElement = getSelf();
+    if (elementNames.empty())
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Element.cpp]: Just a minor note, but you could actually omit this entire early out (including the "if" test"), since the range-based loop below will be skipped if the vector is empty.

File: source/MaterialXCore/Element.h [hunk 11]
```diff
@@ -117,6 +117,14 @@ class Element : public std::enable_shared_from_this<Element>
     ///    the returned path will be relative to this ancestor.
     string getNamePath(ConstElementPtr relativeTo = nullptr) const;
 
+    /// Return the element referred to by the hierarchical path relative
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Element.h]: Perhaps this comment should be updated as well, so that it reads "return the descendant..."?

File: source/MaterialXCore/Element.cpp (hunk 10 above)
Comment (feldstj) [reply source/MaterialXCore/Element.cpp]: Good point.

File: source/MaterialXCore/Element.h (hunk 11 above)
Comment (feldstj) [reply source/MaterialXCore/Element.h]: Ok
